src/util.rs

```diff
@@ -0,0 +1,92 @@
+// Copyright 2021 Contributors to the Parsec project.
+// SPDX-License-Identifier: Apache-2.0
+
+//! Utility code that is shared by multiple subcommands;
+
+use crate::error::{Result, ToolErrorKind};
+use log::{error, info};
+use parsec_client::core::interface::operations::psa_algorithm::{Algorithm, Hash, SignHash};
+use parsec_client::BasicClient;
+use picky_asn1::wrapper::IntegerAsn1;
+use serde::{Deserialize, Serialize};
+use sha2::digest::{Digest, DynDigest};
+
+#[derive(Serialize, Deserialize)]
+struct EccSignature {
+    r: IntegerAsn1,
+    s: IntegerAsn1,
+}
+
+/// Signs a given message using the hashing and signing policy that was associated with the given key when
+/// it was created.
+///
+/// If the signing key allows for the use of any hashing algorithm, then a default hash can optionally be passed
+/// by the caller, and this hash will be used (otherwise the function will fail).
+pub fn sign_message_with_policy(
+    basic_client: &BasicClient,
+    key_name: &str,
+    msg: &[u8],
+    default_hash: Option<Hash>,
```

Comment:
> whether this whole piece of logic belongs in the parsec client itself, rather than the tool, so that any application could take advantage of it, because it is surprisingly verbose to just request a hash-and-sign using the key's own usage policy.

I agree, I think adding something like `fn sign_with_key_alg` in the client would make life a lot easier :) I wonder if we should introduce caching of `Attributes` per key as well, otherwise the client would have to always make an extra request to get that. But maybe this is too advanced for the `BasicClient`!